README.md
# Shutup App - Tell others to be quiet!


## Short summary

Are you tired of asking others to be silent over and over again, when they just forget to be quiet? This HTML5 app uses your microphone to detect if it is too loud in your room. If so, an alarm sound will make clear that the volume is too high.
As "too loud" depends on the user, the limit can be changed. Hint: The alarm sound itself can be louder than the allowed volume. It is not taken into account when calculating the volume, as it is removed by an audio filter.
After a while your colleagues, room mates or whoever disturbs you, will learn to be considerate.

## Demo

You can try the app [here](https://shutup.froso.de). Make sure to use a modern browser supporting the Web Audio Api and the Canvas element.

## Screenshots

![Screenshot1](misc/shutup1.jpeg?raw=true "Screenshot1")
![Screenshot2](misc/shutup2.jpeg?raw=true "Screenshot2")


## List of technologies

* [HTML5 WebAudio Api](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) for audio stuff
* [HTML Canvas](https://developer.mozilla.org/en-US/docs/Glossary/Canvas) to display results
* Build/Compression via [Webpack 2](https://webpack.js.org/)

## How does the app work?

1. It accesses the microphone via the Web Audio Api and starts recording.
```javascript
navigator.mediaDevices.getUserMedia({audio: true}).then(function (stream) {
     // start recording
});

```

2. If the volume exceeds a defined limit, it will activate the alarm.
```javascript

var tooLoud = currentValue > canvasHistorySelectedValue
if (tooLoud) {
     //play alarm
}
```

3. The alarm is generated by a sine oscillator, increasing its volume over time as long as is too loud.

```javascript
var oscillator = audioCtx.createOscillator()
oscillator.type = 'sine'
oscillator.frequency.value = sineFrequency // value in hertz

```

4. The sine tone itself can be louder than the allowed volume, which would normally result in a never ending alarm. But: A sine tone can be removed from the microphone input, if we know the sine frequency (which we do as we generate it by ourselves). In order to do that we use a [biquad notch](https://developer.mozilla.org/en-US/docs/Web/API/BiquadFilterNode) filter.
```javascript

var biquadFilter = audioCtxMic.createBiquadFilter()
biquadFilter.type = "notch"
biquadFilter.frequency.value = sineFrequency
biquadFilter.Q.value = 0.01
```

5. To give visual feedback to the user, we render a canvas with a volume/time graph.
```javascript
// listenToMic, display volume over time on canvas
updateCanvasRegular = function () {

    // update arlam based on the volume
   	// render canvas

     window.requestAnimationFrame(updateCanvasRegular)
 }

```

## Known issues

* Right now I am not too sure about the unit of the volume. It could be "db" but I just multiplied the value with a constant to get a good [0,1] range
* The rendering can be highly optimized
* A UX Designer could improve the UI ;-)

## License

Licensed under MIT

Copyright (c) 2017 [Alexander Thurn](https://github.com/alexanderthurn)


CHANGELOG.md
# Changelog

## 1.3.0
* new algorithm for volume calculation (thinking in steps)

## 1.2.2
* using window.requestAnimationFrame instead of setTimeout for smoother rendering
* using navigator.mediaDevices.getUserMedia  instead of navigator.getUserMedia

## 1.2.1
* Code documented

## 1.2.0
* Readme updated

## 1.1.4
* Simplified UI
* Readme updated
* Better error messages

## 1.1.3
* Crazy colors

## 1.1.2
* Switched to fullscreen view

## 1.1.1
* Readme updated

## 1.1.0
* Added webpack 2
* Code refactoring
* Improved documentation

## 1.0.0
* Initial commit

src/js/helper.js
function getCursorPosition(canvas, event) {
    var rect = canvas.getBoundingClientRect()
    return {x: event.clientX - rect.left, y: event.clientY - rect.top}
}

function getAverageValue(array) {
    var values = 0
    var average

    var length = array.length

    for (var i = 0; i < length; i++) {
        values += array[i]
    }

    average = values / length
    return average
}


function setUserMediaWithFallback() {

// Older browsers might not implement mediaDevices at all, so we set an empty object first
    if (navigator.mediaDevices === undefined) {
        navigator.mediaDevices = {};
    }

// Some browsers partially implement mediaDevices. We can't just assign an object
// with getUserMedia as it would overwrite existing properties.
// Here, we will just add the getUserMedia property if it's missing.
    if (navigator.mediaDevices.getUserMedia === undefined) {
        navigator.mediaDevices.getUserMedia = function (constraints) {

            // First get ahold of the legacy getUserMedia, if present
            var getUserMedia = navigator.getUserMedia ||
                navigator.webkitGetUserMedia ||
                navigator.mozGetUserMedia ||
                navigator.msGetUserMedia

            // Some browsers just don't implement it - return a rejected promise with an error
            // to keep a consistent interface
            if (!getUserMedia) {
                return Promise.reject(new Error('getUserMedia is not implemented in this browser'))
            }

            // Otherwise, wrap the call to the old navigator.getUserMedia with a Promise
            return new Promise(function (resolve, reject) {
                getUserMedia.call(navigator, constraints, resolve, reject)
            });
        }
    }


}

function setRequestAnmationFrameWithFallback() {
    var lastTime = 0;
    var vendors = ['ms', 'moz', 'webkit', 'o'];
    for (var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
        window.requestAnimationFrame = window[vendors[x] + 'RequestAnimationFrame'];
        window.cancelAnimationFrame = window[vendors[x] + 'CancelAnimationFrame']
            || window[vendors[x] + 'CancelRequestAnimationFrame'];
    }

    if (!window.requestAnimationFrame)
        window.requestAnimationFrame = function (callback, element) {
            var currTime = new Date().getTime();
            var timeToCall = Math.max(0, 16 - (currTime - lastTime));
            var id = window.setTimeout(function () {
                    callback(currTime + timeToCall);
                },
                timeToCall);
            lastTime = currTime + timeToCall;
            return id;
        };

    if (!window.cancelAnimationFrame)
        window.cancelAnimationFrame = function (id) {
            clearTimeout(id);
        };
}

function setAudioContextWithFallback() {
    window.AudioContext = window.AudioContext || window.webkitAudioContext || window.mozAudioContext || window.msAudioContext
}


module.exports = {
    getCursorPosition: getCursorPosition,
    setRequestAnmationFrameWithFallback: setRequestAnmationFrameWithFallback,
    getAverageValue: getAverageValue,
    setUserMediaWithFallback: setUserMediaWithFallback,
    setAudioContextWithFallback: setAudioContextWithFallback
}




webpack.config.js
var HtmlWebpackPlugin = require('html-webpack-plugin');
var HtmlWebpackInlineSourcePlugin = require('html-webpack-inline-source-plugin');
var UglifyJsPlugin = require('uglifyjs-webpack-plugin');
var WebpackCleanupPlugin = require('webpack-cleanup-plugin');
var webpack = require('webpack');
var intlJSON = require('./src/res/en.json');
var path = require('path');
var intlJSONStringified = {};
Object.keys(intlJSON).map(function (key) {
    intlJSONStringified['INTL_' + key] = JSON.stringify(intlJSON[key]);
});
intlJSON = intlJSONStringified;

var isProduction = process.env.NODE_ENV === 'production';


console.log('Building with NODE_ENV', process.env.NODE_ENV, path.join(__dirname, "dist"));

var config = {
    /*  watch: !isProduction,
     watchOptions: {
     aggregateTimeout: 300,
     poll: 1000,
     ignored: /node_modules/
     },*/
    devServer: {
        contentBase: path.join(__dirname, "dist"),
        compress: false,
        port: 9000,
        watchContentBase: true
    },
    entry: "./src/js/main.js",
    output: {
        path: __dirname + '/dist/',
        filename: "bundle.js"
    },
    module: {
        loaders: [
            {
                test: /\.html$/,
                loader: 'ejs-loader'
            }
        ]
    },

    plugins: [new WebpackCleanupPlugin()]
};


if (isProduction) {
    config.plugins.push(
        new UglifyJsPlugin({
            compress: {
                sequences: true,
                properties: true,
                drop_debugger: true,
                dead_code: true,
                unsafe: true,
                conditionals: true,
                comparisons: true,
                evaluate: true,
                booleans: true,
                unused: true,
                loops: true,
                cascade: true,
                keep_fargs: false,
                if_return: true,
                join_vars: true,
                drop_console: true
            },
            'mangle-props': true,
            mangle: true,
            beautify: false
        }));
}


config.plugins.push(
    new webpack.DefinePlugin(intlJSON),
    new HtmlWebpackPlugin({
        filename: 'index.html',
        template: "src/index.html",
        minify: {
            minifyJS: true,
            removeEmptyAttributes: true
        },
        inlineSource: '.(js|css)$',
        cache: true
    }));

if (isProduction) {

    config.plugins.push(
        new HtmlWebpackInlineSourcePlugin()
    );
}


module.exports = config;



src/js/view.js
var helper = require('./helper.js')

/* make sure to use https as the web audio api does not like http */

if (location.protocol === 'http:' && location.hostname !== 'localhost' && location.hostname !== '0.0.0.0') {
    location.href = 'https:' + window.location.href.substring(window.location.protocol.length);
}


/* start as soon as things are set up */
document.addEventListener("DOMContentLoaded", function (event) {


    var sineFrequency = 5800  //alarm frequency
    var canvasHistorySelectedValue = 0.5 // the default accepted volume (can be between 0 and 1)
    var arrayLength = Math.floor(document.body.clientWidth)
    var started = true
    var error = false
    var globalAverage
    var canvasHistory = document.getElementById('canvasHistory')
    var btnMute = document.getElementById('btnMute')
    var textInfo = document.getElementById('textInfo');
    var btnInfo = document.getElementById('btnInfo');
    var lineVolume = document.getElementById('lineVolume');
    var axisX = document.getElementById('axisX');
    var axisY = document.getElementById('axisY');
    var ctxHistory = canvasHistory.getContext("2d")
    var time;
    var values = []
    // prefill the array
    for (var i = 0; i < arrayLength; i++) {
        //values.push(0.1+(1+Math.sin(i*0.1))*0.1)
        values.push(0)
    }

    canvasHistory.width = Math.floor(arrayLength)
    canvasHistory.height = Math.floor(document.body.clientHeight)

    // we are using the mute button to show important info for simplicity
    function showMessage(message) {
        btnMute.innerHTML = message
    }

    function showErrorBrowserNotSupporting(message, code) {
        error = true
        btnMute.style.font = "14px Arial"
        btnMute.style.textAlign = 'center'
        btnMute.style.display = 'block';

        if (!message) {
            if (window.location.protocol && window.location.protocol.indexOf('https') > -1) {
                showMessage(INTL_ERROR_SUPPORT_WITH_HTTPS + (code ? '[' + code + ']' : ''))
            } else {
                showMessage(INTL_ERROR_SUPPORT_WITHOUT_HTTPS + (code ? '[' + code + ']' : ''))
            }
        } else {
            showMessage(message)
        }


    }

    // get audio stuff together using helper methods for maximum compatibility
    helper.setAudioContextWithFallback()
    helper.setUserMediaWithFallback()
    helper.setRequestAnmationFrameWithFallback()


    // if the browser does not support audio stuff, show him an error message
    if (!window.AudioContext || !navigator.mediaDevices.getUserMedia || !Array.prototype.slice) {
        showErrorBrowserNotSupporting(null, (!AudioContext * 2) + (!navigator.mediaDevices.getUserMedia * 4) + (!Array.prototype.slice * 8))
        return
    }

    var audioCtx = new window.AudioContext()
    var audioCtxMic = new window.AudioContext()

    // create oscillator node for the alarm
    var oscillator = audioCtx.createOscillator()
    oscillator.type = 'sine'
    oscillator.frequency.value = sineFrequency // value in hertz

    // add a mute filter as we only want to hear the alarm if the volume is too high
    var noiseMuteFilter = audioCtx.createGain()
    noiseMuteFilter.gain.value = 0

    // add another mute filter in case the user does not want any sound at all and only visual effects
    var manualMuteFilter = audioCtx.createGain()
    manualMuteFilter.gain.value = 1

    // wire everything together
    oscillator.connect(noiseMuteFilter)
    noiseMuteFilter.connect(manualMuteFilter)
    manualMuteFilter.connect(audioCtx.destination)

    // start the larm
    oscillator.start()

    // source node is used for mic input
    var sourceNode

    // a biquadfilter is used to remove the sine "volume" from the mic input. otherwise it could result in an endless alarm if the alarmsound itself exceeds the limit
    var biquadFilter = audioCtxMic.createBiquadFilter()
    biquadFilter.type = "notch"
    biquadFilter.frequency.value = sineFrequency
    biquadFilter.Q.value = 0.01

    // analyser and javascript node to process the mic input and check if the volume is too high
    var analyserNode = audioCtxMic.createAnalyser()
    analyserNode.smoothingTimeConstant = 0.3
    analyserNode.fftSize = 1024
    var javascriptNode = audioCtxMic.createScriptProcessor(1024, 1, 1)
    javascriptNode.onaudioprocess = function () {
        var array = new Uint8Array(analyserNode.frequencyBinCount)
        analyserNode.getByteFrequencyData(array)
        var average = helper.getAverageValue(array)
        globalAverage = average * 0.05
    }


    // ask for mic permission
    navigator.mediaDevices.getUserMedia({audio: true}).then(function (stream) {

        //display stuff if it works
        btnInfo.style.display = 'block';
        lineVolume.style.display = 'block';
        textInfo.style.display = 'block';


        console.log('getUserMedia', stream)

        //create the source node using the mic input as a stream
        sourceNode = audioCtxMic.createMediaStreamSource(stream)

        //wire everything together
        sourceNode.connect(biquadFilter)
        biquadFilter.connect(analyserNode)
        analyserNode.connect(javascriptNode)
        javascriptNode.connect(audioCtxMic.destination)


        if (!error) {
            //start rendering and switch title of the mute button as no error occured (and we were reusing the mute button for error messages)
            textInfo.style.display = 'none'
            showMessage(INTL_MUTE_ALARM)
            updateCanvasRegular()
        } else {
            // it did not work, do not show the canvas
            canvasHistory.style.display = 'none'
        }


    }).catch(function (err) {

        // the user did not allow it or it did not work
        error = true;
        console.log(err.name + ": " + err.message);
        showErrorBrowserNotSupporting(null, 16)
        return
    })


    // mute button, setting the gain value to 0 if no volume is wanted
    btnMute.onclick = function () {
        if (!error) {
            if (manualMuteFilter.gain.value > 0) {
                manualMuteFilter.gain.value = 0
                showMessage(INTL_UNMUTE_ALARM)
            } else {
                manualMuteFilter.gain.value = 1
                showMessage(INTL_MUTE_ALARM)
            }
        }
    }

    // display info about the app and axis
    btnInfo.onclick = function () {
        if (textInfo.style.display === 'none') {
            textInfo.style.display = 'block'
            axisX.style.display = 'block';
            axisY.style.display = 'block';
            btnMute.style.display = 'block';
        } else {
            textInfo.style.display = 'none'
            axisX.style.display = 'none';
            axisY.style.display = 'none';
            btnMute.style.display = 'none';
        }
    }

    // changing the maximum volume that is allowed
    canvasHistory.addEventListener('click', function (event) {

        var pos = helper.getCursorPosition(canvasHistory, event)
        canvasHistorySelectedValue = 1 - pos.y / canvasHistory.offsetHeight

    }, false)


    // render canvas
    var updateCanvas = function (options) {

        // use a nice gradient to show which recorded volume was too loud
        var myGradient = ctxHistory.createLinearGradient(0, (1 - canvasHistorySelectedValue) * canvasHistory.height - 1, 0, canvasHistory.height);
        myGradient.addColorStop(0, "darkred");
        myGradient.addColorStop(0.5, "yellow");
        myGradient.addColorStop(1, "green");

        // move the volume line to the correct position
        lineVolume.style.top = (1 - canvasHistorySelectedValue) * canvasHistory.height - 15;


        // render a blue background, but only with 10% of alpha to get a nice effect
        ctxHistory.fillStyle = 'rgba(50,50,255,0.1)'
        ctxHistory.fillRect(0, 0, canvasHistory.width, canvasHistory.height)

        // show shutup text if needed
        ctxHistory.font = Math.min(canvasHistory.height, canvasHistory.width * 0.25) * 0.5 + "px Arial"
        ctxHistory.textAlign = "center"
        ctxHistory.textBaseline = "middle"
        ctxHistory.fillStyle = 'rgba(255,0,0,' + noiseMuteFilter.gain.value * 0.1 + ')'
        ctxHistory.fillText(INTL_SHUTUP_MESSAGE, canvasHistory.width * 0.5 + (0.5 - Math.random()) * canvasHistory.width * 0.01, canvasHistory.height * 0.5 + (0.5 - Math.random()) * canvasHistory.width * 0.01)

        // show some more shutup texts to get a nice effect
        ctxHistory.save()
        ctxHistory.translate(canvasHistory.width * 0.5, canvasHistory.height * 0.5)
        for (var i = 0; i < noiseMuteFilter.gain.value * 10; i++) {
            ctxHistory.rotate(1)
            ctxHistory.scale(0.9, 0.9)
            ctxHistory.fillText(INTL_SHUTUP_MESSAGE, canvasHistory.width * Math.random(), canvasHistory.height * Math.random())
        }
        ctxHistory.restore()

        // render the volume bars
        for (var i = 0; i < arrayLength; i++) {
            var value = values[i]
            ctxHistory.fillStyle = myGradient
            //  ctxHistory.rotate((1.0 / 360.) * Math.PI * 2)
            ctxHistory.fillRect(i * canvasHistory.width / arrayLength, canvasHistory.height - value * canvasHistory.height, canvasHistory.width / arrayLength, value * canvasHistory.height)
        }


    }

    var timeTooLoud, timeTooLoudStart
    // main loop, calls the render method each 30ms + calculates the current average volume + activates the alarm
    var updateCanvasRegular = function () {

        var now = new Date().getTime(),
            dt = now - (time || now);

        var dtFactor = dt / 30.0;
        if (dtFactor < 0.001) {
            dtFactor = 0.001;
        } else if (dtFactor > 5.0) {
            dtFactor = 5.0;
        }
        time = now;

        values.unshift(globalAverage)
        values.pop()

        // we want the current value and not the average of the last 30 recorded audio frames
        var currentValue = globalAverage // helper.getAverageValue(values.slice(0, 10))

        // if the volume is too loud, increase the value of the alarm gain node based on "how loud it is"
        var tooLoud = currentValue > canvasHistorySelectedValue
        if (tooLoud) {


            // get the start of a too loud phase
            if (timeTooLoud === undefined || now - (timeTooLoud || now) > 600) {
                timeTooLoudStart = now
            }


            // increase volume in steps of 1/10 of 3 seconds
            var perc = Math.floor((now - (timeTooLoudStart || now)) / 3000 * 10)

            if (perc > 10) {
                perc = 10;
            }

            if (perc < 2) { // the first two steps are not as loud as the other steps to make the noise not so annoying
                noiseMuteFilter.gain.value = perc * 0.1
            } else {
                noiseMuteFilter.gain.value = perc
            }

            timeTooLoud = now

        } else {

            // no alarm after 600ms of not being too loud
            if (now - (timeTooLoud || now) > 600) {
                noiseMuteFilter.gain.value = 0
            }

        }

        updateCanvas({tooLoud: tooLoud})
        window.requestAnimationFrame(updateCanvasRegular)

    }

// recalculate the canvas size after resize events
    window.onresize = function (event) {
        arrayLength = Math.floor(document.body.clientWidth)
        canvasHistory.width = Math.floor(arrayLength)
        canvasHistory.height = Math.floor(document.body.clientHeight)

        while (values.length < arrayLength) {
            values.push(0)
        }

    };


})
